### PR TITLE
fix(#4036): Section headings within ListBox and Menu are not accessible on mobile screen readers

### LINKED ISSUE #4036: Section headings within ListBox and Menu are not accessible on mobile screen readers
# 🐛 Bug Report 

- Section headings within ListBox and Menu are not accessible on mobile screen readers.
- axe-core throws [Certain ARIA roles must contain particular children](https://dequeuniversity.com/rules/axe/4.6/aria-required-children?application=AxeChrome) error for the Section heading `span` element for both ListBox and Menu, and also for the `span` element with `role="separator"` within a ListBox.

## 🤔 Expected Behavior

- Section headings within ListBox and Menu should be navigable using explore by touch with mobile screen readers.
- axe-core should not throw [Certain ARIA roles must contain particular children](https://dequeuniversity.com/rules/axe/4.6/aria-required-children?application=AxeChrome) error for the Section heading `span` element for both ListBox and Menu, or for the `span` element with `role="separator"` within a ListBox.

## 😯 Current Behavior

- Section headings within ListBox and Menu are not accessible on mobile screen readers.
- axe-core throws [Certain ARIA roles must contain particular children](https://dequeuniversity.com/rules/axe/4.6/aria-required-children?application=AxeChrome) error for the Section heading `span` element for both ListBox and Menu, and also for the `span` element with `role="separator"` within a ListBox.

## 💁 Possible Solution

Return `role="presentation"` instead of `aria-hidden="true"` for `headingProps` in `useListBoxSection` and `useMenuSection`.

Override `role="separator"` with `role="presentation"` for `useListBoxSection`.

## 🔦 Context

Accessibility Audit

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   |  [@adobe/react-spectrum@3.24.1](https://github.com/adobe/react-spectrum/releases/tag/%40adobe%2Freact-spectrum%403.24.1)
| Browser          | Safari iOS / Chrome Android
| Operating System | iOS / Android

## 🧢 Your Company/Team

Adobe/Accessibility

## 🕷 Tracking Issue (optional)

A11Y-5799

### PR DESCRIPTION
**Pull request by majornista**
Closes #4036 

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue 4036](https://github.com/adobe/react-spectrum/issue/4036).
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [x] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

1. Open a ListBox story with sections, like [ListBox with semantic elements (static)](https://reactspectrum.blob.core.windows.net/reactspectrum/ad3ce7ba11a6367e72c6ab5d231a235c3df9bb9b/storybook/iframe.html?providerSwitcher-express=false&scrolling=false&strict=false&providerSwitcher-locale=&providerSwitcher-theme=&providerSwitcher-scale=&args=&id=listbox--with-semantic-elements-static&viewMode=story).
2. Run axe DevTools and verify that the error [Certain ARIA roles must contain particular children](https://dequeuniversity.com/rules/axe/4.6/aria-required-children?application=AxeChrome) is no longer reported for the ListBox.
3. Using an iOS device with VoiceOver running, open a ListBox story with sections, like [ListBox with semantic elements (static)](https://reactspectrum.blob.core.windows.net/reactspectrum/ad3ce7ba11a6367e72c6ab5d231a235c3df9bb9b/storybook/iframe.html?providerSwitcher-express=false&scrolling=false&strict=false&providerSwitcher-locale=&providerSwitcher-theme=&providerSwitcher-scale=&args=&id=listbox--with-semantic-elements-static&viewMode=story).
4. Verify that section headings can be read as text when navigating the ListBox using the VoiceOver cursor.
5. Repeat for stories with sections for [Menu](https://reactspectrum.blob.core.windows.net/reactspectrum/ad3ce7ba11a6367e72c6ab5d231a235c3df9bb9b/storybook/iframe.html?providerSwitcher-express=false&scrolling=false&strict=false&providerSwitcher-locale=&providerSwitcher-theme=&providerSwitcher-scale=&args=&id=menutrigger--isopen), [Picker](https://reactspectrum.blob.core.windows.net/reactspectrum/ad3ce7ba11a6367e72c6ab5d231a235c3df9bb9b/storybook/iframe.html?providerSwitcher-express=false&scrolling=false&strict=false&providerSwitcher-locale=&providerSwitcher-theme=&providerSwitcher-scale=&args=&id=picker--complex-items), and [Combobox](https://reactspectrum.blob.core.windows.net/reactspectrum/ad3ce7ba11a6367e72c6ab5d231a235c3df9bb9b/storybook/iframe.html?providerSwitcher-express=false&scrolling=false&strict=false&providerSwitcher-locale=&providerSwitcher-theme=&providerSwitcher-scale=&args=&id=combobox--sections).

## 🧢 Your Project:

Adobe/Accessibility

### CODE REVIEW

File: packages/@react-aria/listbox/docs/useListBox.mdx [hunk 1]
```diff
@@ -284,25 +284,20 @@ are correct.
...
     heading: section.rendered,
     'aria-label': section['aria-label']
   });
 
-  let {separatorProps} = useSeparator({
-    elementType: 'li'
-  });
-
-  // If the section is not the first, add a separator element.
+  // If the section is not the first, add a separator element to provide visual separatin.
```
Comment (snowystinger) [inline packages/@react-aria/listbox/docs/useListBox.mdx:294]:
```suggestion
  // If the section is not the first, add a separator element to provide visual separation.
```

File: packages/@react-aria/listbox/src/useListBoxSection.ts [hunk 2]
```diff
@@ -47,10 +47,10 @@ export function useListBoxSection(props: AriaListBoxSectionProps): ListBoxSectio
     },
     headingProps: heading ? {
       // Techincally, listbox cannot contain headings according to ARIA.
-      // We hide the heading from assistive technology, and only use it
-      // as a label for the nested group.
+      // We hide the heading from assistive technology, using role="presentation",
+      // and only use it as a label for the nested group.
```
Comment (snowystinger) [inline packages/@react-aria/listbox/src/useListBoxSection.ts:51]: Is this more correct?
```suggestion
      // and only use it as a visual label for the nested group.
```

**Review by snowystinger — COMMENTED** [review]
given how many of these examples there are, should useSeparator be taking some sort of prop to say it should return role=presentation and nothing else?

is elementType: 'li' enough? or are there cases where 'li' would be allowed to have the separator role?

It'd be nice if people implementing just got the right behavior and didn't have to come back to our docs to notice this little change

**Comment by majornista** [comment]
> given how many of these examples there are, should useSeparator be taking some sort of prop to say it should return role=presentation and nothing else?
> 
> is elementType: 'li' enough? or are there cases where 'li' would be allowed to have the separator role?
> 
> It'd be nice if people implementing just got the right behavior and didn't have to come back to our docs to notice this little change

`separator` is allowed within `menu`, which also uses `li`, but is not allowed within `listbox`. Our collections use `group`, so for our purposes, it may make sense to return `role="presentation"` with `elementType: 'li'`, but it is possible to have a `menu` that only uses `separator` to distinguish between sections containing `menuitem`s.

**Review by snowystinger — APPROVED** [review]

**Review by ktabors — APPROVED** [review]
This makes the ListBox/Menu sections accessible on my iPhone. It doesn't identify that it is a group. It looked like ListBoxes can have the role=group, but it didn't add any behavior improvements. Testing on OSX with Safari, the Section behavior is better, but I did notice what seemed like a VO bug where if I navigate one direction it stops on the group without announcing anything, while the other direction it announces the heading and that it's a group.

**Pull request merged**